identity/identity.go

```diff
@@ -103,66 +99,91 @@ func GetIdentityHeader(ctx context.Context) string {
 	return ""
 }
 
-func checkHeader(id *XRHID, w http.ResponseWriter) error {
+var (
+	MissingIdentityError      = errors.New("missing x-rh-identity header")
+	DecodeIdentityError       = errors.New("unable to b64 decode x-rh-identity header")
+	InvalidOrgIdIdentityError = errors.New("x-rh-identity header has an invalid or missing org_id")
+	MissingIdentityTypeError  = errors.New("x-rh-identity header is missing type")
+)
```

Comment:
The package now export and return these errors. The middleware still writes them into body with HTTP 400 code to keep the old behavior but callers can implement their own middleware with the newly introduced exported functions.

Reply:
I don't have any hard evidence to back this up, but the pattern I have observed throughout the Go stdlib is that when naming an error type (i.e. in the type declaration), the pattern `FooError` is used:

```go
type IdentityError struct {
    description string
    wrappedError error
}

func (e *IdentityError) Error() string {
    var s string
    if e.wrappedError != nil {
        s = fmt.Sprintf("%w", e.wrappedError)
    }
    if description != "" {
        s += ": " + description
    }
    return s
}
```

Then, when declaring a variable of that type, the pattern `ErrSpecificDescription` is used:

```go
var ErrMissingIdentityHeader = IdentityError{description: "missing x-rh-identity header"}
```

This would allow users to compare errors using the `errors.Is` and `errors.Wraps` API. I don't think that API can be used directly with `errors.errorString` (the type returned by `errors.New`.

Reply:
Up until now, I have been mixing those. This make sense, I like that and I will start enforcing this on my projects too.